keps/sig-scheduling/2044-topology-awareness-in-kube-scheduler/README.md

```diff
@@ -0,0 +1,767 @@
+    Parent     string           `json:"parent,omitempty"`
+    Costs      []CostInfo       `json:"costs,omitempty"`
+    Attributes []AttributeInfo  `json:"attributes,omitempty"`
+    Resources  []ResourceInfo   `json:"resources,omitempty"`
+}
+
+type ResourceInfo struct {
+    Name        string `json:"name"`
+    Allocatable string `json:"allocatable"`
+    Capacity    string `json:"capacity"`
+}
+
+type CostInfo struct {
+    Name  string `json:"name"`
+    Value int	 `json:"value"`
+}
+
+type AttributeInfo struct {
+    Name  string `json:"name"`
+    Value string `json:"value"`
+}
+// Kubelet writes to NodeResourceTopology
+// and scheduler plugin reads from it 
+// Real world example of how these fields are populated is as follows:
+// Cells:
+//   Name:     node-1
+//   Type:     Node
+//   Costs:
+//     Name:   node-0
+//     Value:  20
+//     Name:   node-1
+//     Value:  10
+//   Attributes:
+//    Name: performance-profile
+//    Value: high-performance-profile
+//   Resources:
+//     Name:         example.com/deviceB
+//     Allocatable:  2
+//     Capacity:     2
+//     Name:         example.com/deviceA
+//     Allocatable:  2
+//     Capacity:     2
+//     Name:         cpu
+//     Allocatable:  4
+//     Capacity:     4
+
+```
+
+
+### Plugin implementation details
+
+### Description of the Algorithm
+
+The algorithm which of the scheduler plugin is as follows:
+
+1. At the filter extension point of the plugin, the QoS class of the pod is determined, in case it is a best effort pod or the 
+   Topology Manager Policy configured on the node is not single-numa-node, the node is not considered for scheduling
+1. The Topology Manager Scope is determined.
+1. While interating through the containers of a pod
+	* A bitmask is created where each bit corresponds to a NUMA cell and are all the bits are set. If the resources cannot be aligned on the NUMA cell,
+	the bit should be unset.
+	* For each resource requested in a container, a new resourceBitmask is created to determined which NUMA cell is a good fit for each resource
+		1. If requested resource cannot be found on a node, it is unset as available NUMA cell
+		1. If an unknown resource has 0 quantity, the NUMA cell should be left set.
+	* The following checks are performed:
+		1. Add NUMA cell to the resourceBitmask if resource is cpu and it's not guaranteed QoS, since cpu will flow
+		1. Add NUMA cell to the resourceBitmask if resource is memory and it's not guaranteed QoS, since memory will flow
+		1. Add NUMA cell to the resourceBitmask if zero quantity for non existing resource was requested
+		1. otherwise check amount of resources
+	* Once the resourceBitMark is determined it is ANDed with the cummulative bitmask
+4. If resources cannot be aligned from the same NUMA cell for a container, alignment cannot be achieved for the entire pod and the resource cannot be 
+   aligned in case of the pod under consideration. Such a pod is returned with a Status Unschedulable 
```

Comment:
If we return `Unschedulable`, it will enter the postfilter phase. But we don't know which cell the pod is assigned to. So the DeletePod in PostFilter will not change the `NodeResourceTopology`. So it will always return false in `PodPassesFiltersOnNode`.  Then this pod can never choose one or more pods to seize preempt.